fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveHaving.java

```diff
@@ -0,0 +1,184 @@
+                return createPlan(having, resolver);
+            })
+        );
+    }
+
+    static class Resolver {
+
+        private final List<NamedExpression> outputExpressions;
+        private final List<Expression> groupByExpressions;
+        private final Map<Expression, Slot> substitution = Maps.newHashMap();
+        private final List<NamedExpression> newOutputSlots = Lists.newArrayList();
+
+        Resolver(LogicalAggregate<? extends Plan> aggregate) {
+            outputExpressions = aggregate.getOutputExpressions();
+            groupByExpressions = aggregate.getGroupByExpressions();
+        }
+
+        public void resolve(Expression expression) {
+            Pair<Optional<Expression>, Boolean> result = lookUp(expression);
+            Optional<Expression> found = result.first;
+            boolean isFoundInOutputExpressions = result.second;
+
+            if (found.isPresent()) {
+                if (found.get() instanceof NamedExpression) {
+                    substitution.put(expression, ((NamedExpression) found.get()).toSlot());
+                    if (!isFoundInOutputExpressions) {
+                        newOutputSlots.add(((NamedExpression) found.get()).toSlot());
+                    }
+                } else {
+                    // This expression was found in group-by expressions. E.g. group by (a + 1) having (a + 1)
+                    generateAliasForNewOutputSlots(expression);
+                }
+            } else {
+                if (expression instanceof SlotReference) {
+                    throw new AnalysisException(expression.toSql() + " in having clause should be grouped by.");
+                } else if (expression instanceof AggregateFunction) {
+                    if (checkWhetherNestedAggregateFunctionsExist((AggregateFunction) expression)) {
+                        throw new AnalysisException("Aggregate functions in having clause can't be nested: "
+                                + expression.toSql() + ".");
+                    }
+                    generateAliasForNewOutputSlots(expression);
+                } else {
+                    for (Expression child : expression.children()) {
+                        resolve(child);
+                    }
+                }
+            }
+        }
+
+        private Pair<Optional<Expression>, Boolean> lookUp(Expression expression) {
+            Optional<Expression> found = outputExpressions.stream()
+                    .filter(source -> isEquivalent(source, expression))
+                    .map(source -> (Expression) source)
+                    .findFirst();
+            if (found.isPresent()) {
+                return Pair.of(found, true);
+            }
+
+            found = groupByExpressions.stream().filter(source -> isEquivalent(source, expression)).findFirst();
+            return Pair.of(found, false);
+        }
+
+        private boolean isEquivalent(Expression source, Expression expression) {
```

Comment:
add some comment to explain the meaning of equality

Reply:
Added.